### ISSUE TITLE
dpkg adds extra / in Filename

### ISSUE THREAD

**Issue by bedge** [Homebrew/homebrew#44053]
%> dpkg --version
Debian 'dpkg' package management program version 1.18.2 (darwin-amd64).

Short version:
Ubuntu:
```
Filename: binary/nim-cmas_3.49.0-SNAPSHOT-1_all.deb
```
OSX:
```
Filename: binary//nim-cmas_3.49.0-SNAPSHOT-1_all.deb
```
Note the extra /, binary// instead of binary/

tl;dr

Here's my dir:
```
%> tree binary
binary
├── nim-cmas_3.49.0-SNAPSHOT-1_all.deb
└── override
```
Running on ubuntu shows:
```
0 %> dpkg-scanpackages binary binary/override
Package: nim-cmas
Source: nim-cmas
Version: 3.49.0-SNAPSHOT-1
Architecture: all
Maintainer: bedge
Installed-Size: 37247
Provides: nim-cmas
Filename: binary/nim-cmas_3.49.0-SNAPSHOT-1_all.deb
Size: 34293426
MD5sum: 3d0518178d44e6419159b46e2f2cbf7a
SHA1: 877b07db9cbf6b893551c9885526fd9650091195
SHA256: 8ea27a44d15955f1cf84d35871c0044d9e2d175f9c22684bff716bc95b82de29
Section: java
Priority: optional
Description: nim-cmas

dpkg-scanpackages: warning: Packages in archive but missing from override file:
dpkg-scanpackages: warning:   nim-cmas
dpkg-scanpackages: info: Wrote 1 entries to output Packages file.

%> dpkg-scanpackages --version
Debian dpkg-scanpackages version 1.16.12.

```

Running on osx shows:
```
%> dpkg-scanpackages binary binary/override
Package: nim-cmas
Source: nim-cmas
Version: 3.49.0-SNAPSHOT-1
Architecture: all
Maintainer: bedge
Installed-Size: 37247
Provides: nim-cmas
Filename: binary//nim-cmas_3.49.0-SNAPSHOT-1_all.deb
Size: 34293426
MD5sum: 3d0518178d44e6419159b46e2f2cbf7a
SHA1: 877b07db9cbf6b893551c9885526fd9650091195
SHA256: 8ea27a44d15955f1cf84d35871c0044d9e2d175f9c22684bff716bc95b82de29
Section: java
Priority: optional
Description: nim-cmas

dpkg-scanpackages: warning: Packages in archive but missing from override file:
dpkg-scanpackages: warning:   nim-cmas
dpkg-scanpackages: info: Wrote 1 entries to output Packages file.

%> dpkg-scanpackages --version
Debian dpkg-scanpackages version 1.18.2.
```

**Comment by apjanke** [comment]
Upstream issue: compatibility with BSD-style `find` command.

Looks like this is a difference in behavior between the BSD and GNU versions of `find`. Linux comes with GNU `find`; OS X comes with BSD `find`.

The `dpkg-scanpackages` script calls the external `find` command, and puts a trailing slash on its directory argument.

```
open my $find_h, '-|', 'find', '-L', "$binarydir/", @find_args, '-print'
     or syserr(g_("couldn't open %s for reading"), $binarydir);
```

Looks like the GNU version of `find` will canonicalize the filenames it outputs by eliding repeated slashes after the input argument. The BSD `find` won't. Here's a comparison between the BSD-style OS X system `find` and GNU `find` as `gfind` from `brew install findutils`. 

```
$ find -L blah/ -print
blah/
blah//test.deb
$ gfind -L blah/ -print
blah/
blah/test.deb
```

I'd bet other occurrences of this are caused by similar `find` invocations.

This is probably just a cosmetic issue, because in Unix file paths, repeated slashes mean the same thing as a single slash.

Upstream could fix this by adding compatibility with BSD `find` behavior by canonicalizing the filenames it gets from `find` output. Or switch to a pure-perl implementation using `File::Find`. Or whatever.

**Comment by apjanke** [comment]
Looks like doing an `ENV["FIND"]` would also require an upstream change to support it. Their autoconf setup (in `m4/dpkg-progs.m4`) is defined to detect a GNU tar, but not a find.

**Comment by bedge** [comment]
It's not just cosmetic. I'm using this to populate s3 repos that I get via http, then the // matters:
```
Err http://s3.amazonaws.com/xxxx-xxxx-xxxx/artifacts/bedge-dc-16/cmas/ binary/ nim-cmas 3.49.0-SNAPSHOT-1
  403  Forbidden
E: Failed to fetch http://s3.amazonaws.com/xxxx-xxxx-xxxx/bedge/artifacts/bedge-dc-16/cmas/binary//nim-cmas_3.49.0-SNAPSHOT-1_all.deb  403  Forbidden
```
Maybe looking for gfind in the path? Can one make the assumption that gfind will always be gnu find?

One could hack this by using
```
sudo ln -s /usr/local/bin/gfind  /usr/local/bin/find
```
but not a great option for a large team.

**Comment by apjanke** [comment]
"Looking for gfind in the path" would require that upstream change to detect a GNU `find`. That's what they do for detecting a GNU `tar`: check for `tar`, `gtar`, and `gnutar` on the path and check each for GNU-ish behavior.

Yeah, that's a lousy hack linking `find`.

For yourself, you could do a fix on your side by simply canonicalizing each file name before including them in the URLs (where doubled `//` are significant). That's probably the quickest and most robust thing for you: would require no system-local configuration or waiting on upstream. What language are you doing that URL construction in?

**Comment by bedge** [comment]
I'm good for now using this:
```
	dpkg-scanpackages binary binary/override | sed -e "s,//,/," | gzip -9c > binary/Packages.gz
```
as it'll become a noop when it's fixed.

**Comment by apjanke** [comment]
You should probably [report this as an upstream bug](https://www.debian.org/doc/manuals/debian-bugs/ch2.en.html), in any case. Since they're willing to handle systems with non-GNU `tar`s, they'd probably be willing to handle this, too.

Package: dpkg
Version: 1.18.2

There's a similar bug at https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=122255 which you could reference.

**Comment by DomT4** [comment]
Please do report this upstream. The `dpkg` team have been very receptive towards cross-platform fixes, and I suspect they'd be willing to work something here as well.

**Comment by bedge** [comment]
I finally got around to filing an upstream bug. Looks like it's already fixed in 18.4:
https://anonscm.debian.org/cgit/dpkg/dpkg.git/diff/?id=3f8099b

**Comment by DomT4** [comment]
`1.18.4` was merged, so presuming resolved.
